+nosnoc/+time_freezing/cls_inelastic.m

```diff
@@ -0,0 +1,189 @@
+function [pss_model] = cls_inelastic(cls_model, opts)
+    import casadi.*
+    dims = cls_model.dims;
+    pss_model = nosnoc.model.Pss();
+    pss_model.dims = dims;
+
+    % transfer all common properties to pss_model
+    common_props = properties('nosnoc.model.Base');
+    for ii=1:length(common_props)
+        prop = common_props{ii};
+        pss_model.(prop) = cls_model.(prop);
+    end
+    
+    dims.n_contacts = length(cls_model.f_c);
+    % check dimensions of contacts
+    if isempty(dims.n_dim_contact)
+        warning('nosnoc: Please n_dim_contact, dimension of tangent space at contact (1, 2 or 3)')
+        dims.n_dim_contact = 2;
+    end
+    
+    % quadrature state
+    dims.n_quad = 0;
+    f_quad = [];
+    if opts.time_freezing_quadrature_state
+        % define quadrature state
+        L = define_casadi_symbolic(opts.casadi_symbolic_mode,'L',1);
+        pss_model.lbx = [pss_model.lbx;-inf];
+        pss_model.ubx = [pss_model.ubx;inf];
+        pss_model.x = [pss_model.x;L];
+        pss_model.x0 = [pss_model.x0;0];
+        f_quad = pss_model.f_q;
+        pss_model.f_q = 0;
+        if ~isempty(pss_model.f_q_T)
+            pss_model.f_q_T  = pss_model.f_q_T + L;
+        else
+            pss_model.f_q_T = L;
+        end
+        dims.n_quad = 1;
+    end
+    
+    % uneven number of states = it is assumed that the clock state is defined.
+    t = define_casadi_symbolic(opts.casadi_symbolic_mode,'t',1);
+    % update lower and upper bounds of lbx and ubx
+    pss_model.lbx = [pss_model.lbx;-inf];
+    pss_model.ubx = [pss_model.ubx;inf];
+    pss_model.x = [pss_model.x;t];
+    pss_model.x0 = [pss_model.x0;0];
+    
+    % normal and tangential velocities
+    eps_t = 1e-7;
+    v_normal = cls_model.J_normal'*cls_model.v;
+    if cls_model.friction_exists
+        if dims.n_dim_contact == 2
+            v_tangent = (cls_model.J_tangent'*cls_model.v)';
```

Comment:
why are the tangential velocities of 2 and 3 treated differently?
For example, there might be few discs in 2d plane, and there are multiple possible tangential velocities (and more might be active at time, e.g., one discs pushed a middle disc, the middle disch pushes the last one, to get it to its goal (a nice ocp example, where the contact between the very first (actuated) and last is not modeled)

Reply:
You mean 1 and 2? I don't quite understand the question.